zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/MultipleAddresses.java

```diff
@@ -0,0 +1,226 @@
+    }
+
+    public void addAddress(InetSocketAddress address) {
+        addresses.add(address);
+    }
+
+    /**
+     * Returns reachable address. If none is reachable than throws exception.
+     *
+     * @return address which is reachable.
+     * @throws NoRouteToHostException if none address is reachable
+     */
+    public InetSocketAddress getReachableAddress() throws NoRouteToHostException {
+        AtomicReference<InetSocketAddress> address = new AtomicReference<>(null);
+        getInetSocketAddressStream().forEach(addr -> checkIfAddressIsReachableAndSet(addr, address));
+
+        if (address.get() != null) {
+            return address.get();
+        } else {
+            throw new NoRouteToHostException("No valid address among " + addresses);
+        }
+    }
+
+    /**
+     * Returns reachable address or first one, if none is reachable.
+     *
+     * @return address which is reachable or fist one.
+     */
+    public InetSocketAddress getReachableOrOne() {
+        InetSocketAddress address;
+        try {
+            address = getReachableAddress();
+        } catch (NoRouteToHostException e) {
+            address = getOne();
+        }
+        return address;
+    }
+
+    /**
+     * Performs a DNS lookup for addresses.
+     *
+     * If the DNS lookup fails, than address remain unmodified.
+     */
+    public void recreateSocketAddresses() {
+        Set<InetSocketAddress> temp = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        temp.addAll(getInetSocketAddressStream().map(this::recreateSocketAddress).collect(Collectors.toSet()));
+        addresses = temp;
+    }
+
+    /**
+     * Returns first address from set.
+     *
+     * @return address from a set.
+     */
+    public InetSocketAddress getOne() {
+        return addresses.iterator().next();
+    }
+
+    private void checkIfAddressIsReachableAndSet(InetSocketAddress address,
+                                                 AtomicReference<InetSocketAddress> reachableAddress) {
+        for (int i = 0; i < 5 && reachableAddress.get() == null; i++) {
+            try {
+                if (address.getAddress().isReachable((i + 1) * timeout)) {
+                    reachableAddress.compareAndSet(null, address);
+                    break;
+                }
+                Thread.sleep(timeout);
+            } catch (NullPointerException | IOException | InterruptedException ignored) {
+            }
+        }
+    }
+
+    private InetSocketAddress recreateSocketAddress(InetSocketAddress address) {
+        try {
+            return new InetSocketAddress(InetAddress.getByName(address.getHostString()), address.getPort());
+        } catch (UnknownHostException e) {
+            return address;
+        }
+    }
+
+    private Stream<InetSocketAddress> getInetSocketAddressStream() {
+        if (addresses.size() > 1) {
+            return addresses.parallelStream();
```

Comment:
Can we ad a comment for this parallelStream ?
is it really needed ?

Reply:
sure, I will add a comment.

Basically the idea is, that using parallelStream we can do hostname resolutions or isReachable checks parallel on all addresses. I think this is a meaningful optimization, otherwise we can loose many seconds e.g. during a leader election when the single working address is in the end of the list.

Although I don't think this function is really necessary. I think we can simply use parallelStreams which would work for sets with single element too.